File: src/rectangle-overlap/rectangle-overlap.js
/**********************************************************************************************************************
 * Write a function to find the rectangular intersection of two given love rectangles.
 * 
 * As with the example above, love rectangles are always "straight" and never "diagonal."
 * More rigorously: each side is parallel with either the x-axis or the y-axis.
 */
function findRectangleOverlap( rectangle1, rectangle2 )
{
	var xOverlap = findRangeOverlap( rectangle1.x, rectangle1.width, rectangle2.x, rectangle2.width );
	var yOverlap = findRangeOverlap( rectangle1.y, rectangle1.height, rectangle2.y, rectangle2.height );

	if ( !xOverlap || !yOverlap ) {
		console.log( 'No intersection' );
		return undefined;
	}
	else {
		var overlapRect = {
			x: xOverlap[0],
			y: yOverlap[0],
			width: xOverlap[1],
			height: yOverlap[1]
		};
		return overlapRect;
	}
}


/**********************************************************************************************************************
 * Helper function to compute x and y overlap independantly
 */
function findRangeOverlap( point1, range1, point2, range2 )
{
	var farStartPoint = Math.max( point1, point2 );
	var nearEndPoint = Math.max( point1 + range1, point2 + range2 );
	var overlap = nearEndPoint - farStartPoint;
	
	if ( overlap < 0 ) {
		return [undefined, undefined];
	}
	else {
		return [farStartPoint, overlap];
	}
}

File: src/rectangle-overlap/rectangle_overlap.py
def find_rectangle_overlap( rec1, rec2 ):
	"""
	**********************************************************************************************************************
	Write a function to find the rectangular intersection of two given love rectangles.

	As with the example above, love rectangles are always "straight" and never "diagonal."
	More rigorously: each side is parallel with either the x-axis or the y-axis.
	"""
	x_overlap = find_range_overlap( rec1.x, rec1.width, rec2.x, rec2.width )
	y_overlap = find_range_overlap( rec1.y, rec1.height, rec2.y, rec2.height )

	if !x_overlap[0] or !y_overlap[0]:
		return None

	return {
		'x': x_overlap[0],
		'y': y_overlap[0],
		'width': x_overlap[1],
		'height': y_overlap[1]
	}


def find_range_overlap( point1, range1, point2, range2 ):
	"""
	**********************************************************************************************************************
	Helper function to find the overlap in one dimension
	"""
	## all we need to know is if the right starting point is left of the left ending point ##
	right_start = max( point1, point2 )
	left_end = min( point1 + range1, point2 + range2 )

	if right_start > left_end:
		return [ None, None ]

	else:
		return [ right_start, left_end - right_start ]
